### ISSUE TITLE
UI sees old and new state of a re-run flow

### ISSUE THREAD

**Issue by hjoliver** [cylc/cylc-ui#556]
**Describe the bug**

If a stop a flow, the tree view remembers its final state.

If I then run the flow again after reconfiguring it to look different, the tree view still shows the old stopped state along with the new restarted state

**Release version(s) and/or repository branch(es) affected?**

Current master

**Steps to reproduce the bug**

```ini
[scheduling]                                                                    
    initial cycle point = now                                                  
    [[graph]]                                                                   
        R1 = foo => bar & baz & qux                                             
[runtime]                                                                       
    [[foo]]                                                                     
       script = "sleep 20"        
```
Run the above flow, stop it while the first task is running, comment out `initial cycle point = now` to make it an integer flow instead of date-time, then run it again.  

**Expected behavior**

Clear out the old state, once the flow starts running again.

**Screenshots**

![Screenshot from 2020-12-09 12-24-10](https://user-images.githubusercontent.com/2362137/101553902-73b90c80-3a1a-11eb-8f6c-4aabbef68f8b.png)

**Pull requests welcome!**
This is an Open Source project - please consider contributing a bug fix
yourself (please read `CONTRIBUTING.md` before starting any work though).

**Comment by kinow** [comment]
Reproduced successfully. Thanks for sharing a workflow :+1:

**Comment by kinow** [comment]
Ah! Interesting @hjoliver . I've tracked down the problem.

I thought it was strange that we are seeing this last state, since we have `deltas.shutdown`. That `shutdown` is a boolean, that the backend sends to inform the UI a workflow is done.

When we receive it, we clear the `CylcTree` object attached to the Workflow (and Tree) view. Each Tree component added via lumino widgets get a reference for this same `CylcTree`.

So the lumino Tree component was supposed to be empty. The reason why it is not, is that we have now:

- Initial data burst delta, when the workflow starts (`CylcTree` object is created here)
- Several deltas, when workflow has nodes added/removed/updated (`CylcTree` object is updated)
- One final delta with `shutdown = true` (`CylcTree` is cleared with this message)
- Yet another message, with the state of the workflow, and `shutdown = false`.

![Untitled](https://user-images.githubusercontent.com/304786/101557347-b9c59e80-3a21-11eb-9e19-07fb1744dbe4.png)

If you run a delta subscription in GraphiQL for a workflow, and you include the `deltas.shutdown`, you should see it blink when you `cylc stop 556`, and temporarily display `shutdown = true`. The UI does remove everything, but it's so quick it's hard to see.

@hjoliver I think I know how to clear the tree when `shutdown=true` is received, and prevent it from re-creating the workflow. But that will take a little time to implement, but much longer to test and write tests.

I **think** it could be fixed too if the last message wasn't received after the `shutdown=true` was sent, and I **guess** it could be fixed with the event-driven data store that @dwsutherland is working on. What do you think?

**Comment by kinow** [comment]
@dwsutherland if you have some time to take a look at this later (I'm in-between pull requests, just triaging issues, no hurry on this one) :point_up:

**Comment by kinow** [comment]
I thought the last message, the one after `{ shutdown: true }` had only the `ID`. So I quickly wrote some code that ignores that message; but in reality this extra message appears to contain the same thing as the message with `shutdown: true`. So there goes my plan to tackle this issue in the UI :sweat_smile: 

![image](https://user-images.githubusercontent.com/304786/104965476-aefc2180-5a43-11eb-9bc9-b74e7c2dfc29.png)

(see all the data under `workflow.added`)

**Comment by dwsutherland** [comment]
I'll have to have a look, but the last one might be the delta from the UIS when it transitions from running => installed (?)  (and the UIS hasn't purged it's data-store yet, might be intentional as the last known state).

So, that can be changed at the UIS.. You could treat and `added.workflow` as replace the previous data set (as it only happens on these transitions or restart/reload/run)

**Comment by kinow** [comment]
>You could treat and added.workflow as replace the previous data set (as it only happens on these transitions or restart/reload/run)

I will have to think about this one. Because I think I get `added.workflow` multiple times for the same workflow. Probably because we have `taskProxies` and `jobs` changing within the `added.workflow`? See screencast below.

![GIFrecord_2021-01-26_122440](https://user-images.githubusercontent.com/304786/105778660-aa021980-5fd1-11eb-92e6-3bb83c0cfb37.gif)

So not sure how I can tell the `delta.added.workflow` is for a started/restarted/reloaded workflow. Should I look if the `delta.added.workflow` attributes (`state`, `name`, `port`, etc) changed, ignoring `taskProxies`, `jobs`, cycle points, etc?

**Comment by dwsutherland** [comment]
Looks like they are in the wrong place... Will need to fix that..

Additions of jobs to workflow should be in `delta.updated.workflow`

**Comment by kinow** [comment]
> Looks like they are in the wrong place... Will need to fix that..
> 
> Additions of jobs to workflow should be in `delta.updated.workflow`

Hmmm, that will require some changes to how the UI handles deltas I think.

@dwsutherland what if I had

```graphql
subscription {
deltas {
  added {
    workflow { ... }
    familyProxies : taskProxies { firstParent {} ... }
    taskProxies { firstParent {} ... }
    jobs { firstParent {} ... }
  }
}
}
```

![GIFrecord_2021-01-29_113317](https://user-images.githubusercontent.com/304786/106206988-e463f480-6225-11eb-88ea-d9ef499cf55a.gif)

Then `deltas.added.workflow` would mean a newly added workflow. And it wouldn't change again unless the state/port/host/etc changed.

And any task/job/familyproxy/etc in `deltas.added.taskProxies` or `...added.jobs` would mean a new task or job must be added to the tree. The `deltas.updated.taskProxies` and `...updated.jobs` would mean that something changed in a node of the tree.

**Comment by kinow** [comment]
I still would have issues with the extra message sent after the message with `shutdown: true` though :disappointed_relieved:

**Comment by dwsutherland** [comment]
Yes, the intension is to use `delta.added.<node>` for new nodes (jobs/proxies ...etc)
(and `delta.added.workflow` for new workflow)

**Comment by dwsutherland** [comment]
Also, you can just not include id under `delta.added.workflow` as it's already under `delta` .. However, it may not show up like that once I fix the job addition causing this issue.

**Comment by kinow** [comment]
> Also, you can just not include id under `delta.added.workflow` as it's already under `delta` .. However, it may not show up like that once I fix the job addition causing this issue.

Brilliant, I can move it out of `workflow`, that shouldn't be too difficult. I won't work on this for another week, so once I'm back will check again with you to see if the job addition issue has been fixed or not. If it has been fixed, I'll sync code and see what the UI looks like, etc.

Thanks @dwsutherland !

**Comment by dwsutherland** [comment]
Looking into this.. as it's definitely strange that the added.workflow.jobs gets populated when new ones are added:
```
subscription {
  deltas (workflows: ["sutherlander|baz/run1"], stripNull: true) {
    added {
      workflow {
        id
        stateTotals
        jobs (deltaStore: true, deltaType: "added") {
          id
          state
        }
        stateTotals
      }
    }
  }
}
```
(this cuts it down to just the new updates... but still)

@kinow - this should stop you from switching to the more flat approach

**Comment by kinow** [comment]
#542 is using a flat structure, and as note to self, I think I fixed the issue in that branch, but won't include the change to simplify the review. Here's the diff:

```diff
diff --git a/src/components/cylc/tree/deltas.js b/src/components/cylc/tree/deltas.js
index e51a0d64..0fb6fb78 100644
--- a/src/components/cylc/tree/deltas.js
+++ b/src/components/cylc/tree/deltas.js
@@ -16,6 +16,7 @@
  */
 
 import isArray from 'lodash/isArray'
+import WorkflowState from '@/model/WorkflowState.model'
 import {
   createWorkflowNode,
   createCyclePointNode,
@@ -244,6 +245,14 @@ export default function (data, workflow, lookup, options) {
       }
     }
   }
+  // In case we have a workflow that is stopped and contains data. When we
+  // receive a delta when the workflow has changed status, we must reset the
+  // tree structure before applying new deltas.
+  if (deltas.added && deltas.added.workflow && deltas.added.workflow.status) {
+    if (deltas.added.workflow.status !== WorkflowState.STOPPED.name) {
+      CylcTree.clear(workflow)
+    }
+  }
   // the tree was created, and now the next messages should contain
   // 1. data added in deltas.added
   // 2. data updated in deltas.updated
```

(the GScan PR includes some refactorings that simplified and decoupled deltas and node-creation code, hence the simple diff :neckbeard: )
